flink-ml-lib/src/main/java/org/apache/flink/ml/common/broadcast/operator/CacheStreamOperator.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import org.apache.commons.collections.IteratorUtils;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+/** The operator that process all broadcast inputs and stores them in {@link BroadcastContext}. */
+public class CacheStreamOperator<OUT> extends AbstractStreamOperatorV2<OUT>
+        implements MultipleInputStreamOperator<OUT>, BoundedMultiInput, Serializable {
+    /** names of the broadcast DataStreams. */
+    private final String[] broadcastNames;
+    /** input list of the multi-input operator. */
+    private final List<Input> inputList;
+    /** output types of input DataStreams. */
+    private final TypeInformation<?>[] inTypes;
+    /** caches of the broadcast inputs. */
+    private final List<?>[] caches;
+    /** state storage of the broadcast inputs. */
+    private ListState<?>[] cacheStates;
+    /** cacheReady state storage of the broadcast inputs. */
+    private ListState<Boolean>[] cacheReadyStates;
+
+    public CacheStreamOperator(
+            StreamOperatorParameters<OUT> parameters,
+            String[] broadcastNames,
+            TypeInformation<?>[] inTypes) {
+        super(parameters, broadcastNames.length);
+        this.broadcastNames = broadcastNames;
+        this.inTypes = inTypes;
+        this.caches = new List[inTypes.length];
+        for (int i = 0; i < inTypes.length; i++) {
+            caches[i] = new ArrayList<>();
+        }
+        this.cacheStates = new ListState[inTypes.length];
+        this.cacheReadyStates = new ListState[inTypes.length];
+
+        inputList = new ArrayList<>();
+        for (int i = 0; i < inTypes.length; i++) {
+            inputList.add(new ProxyInput(this, i + 1));
+        }
+    }
+
+    @Override
+    public List<Input> getInputs() {
+        return inputList;
+    }
+
+    @Override
+    public void endInput(int i) {
+        BroadcastContext.markCacheFinished(
+                Tuple2.of(broadcastNames[i - 1], getRuntimeContext().getIndexOfThisSubtask()));
+    }
+
+    @Override
+    public void snapshotState(StateSnapshotContext context) throws Exception {
+        super.snapshotState(context);
+        for (int i = 0; i < inTypes.length; i++) {
+            cacheStates[i].clear();
+            cacheStates[i].addAll((List) caches[i]);
+            cacheReadyStates[i].clear();
+            boolean isCacheFinished =
+                    BroadcastContext.isCacheFinished(
+                            Tuple2.of(
+                                    broadcastNames[i],
+                                    getRuntimeContext().getIndexOfThisSubtask()));
+            cacheReadyStates[i].add(isCacheFinished);
+        }
+    }
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
+        for (int i = 0; i < inTypes.length; i++) {
+            cacheStates[i] =
```

Comment:
There is an utility  `OperatorStateUtils.getUniqueElement` to get the unique list element.

Reply:
Thanks Yun. I have made the changes.